### PR TITLE
Larger refactor of addon structure, by Oscarwika

### PR DESCRIPTION
**Pull request by jobackman**
Big thanks to @oscarwika for this! https://github.com/oscarwika/SpellNotifications

- Now has sounds!!
  - Will play a _buzz_ when your pet dies
- Code is a lot more readable and maintanable

Outside of @oscarwika contributions I made a few adjustments (no functional changes to addon users): 

- Many of the functions defined were scoped to the Global object. I'd like to avoid cluttering the global scope, so that we dont run into any namespace collisions. So I now put them all into a single table `SpellNoficiations` that is available on the global scope. 
  - So now we instead call `addon.Colors()` or `addon.Messages()` from within the addon
  - Or `SpellNotifications.OnEvent()`, `SpellNotifications.Messages()`, `SpellNotifications.print()` from outside the addon scope
- Re-used the `Colors` table in all the calls to `addon.print`, that way we dont need to do any extra lower-case checking for the strings sent
- New data table for the sizes we can print in, also used with all print calls
- A new `Affiliations` data table (perhaps there's a better name for it) for the combat log flag constants that are used to check the source and destination of the events
- Instead of a function that checks if events match and if the combat log flags are as expected, I made a new table of functions to be used with the Affiliation table:

```lua
-- There could be a better name than `cast.by`, `cast.on` etc. 
-- I just wanted it to be readable as viewed down below
local cast = {}
function cast.by(affiliation)
	return bit.band(sourceFlags, affiliation) > 0
end
function cast.on(affiliation)
	return bit.band(destFlags, affiliation) > 0
end
function cast.notOn(affiliation)
	return bit.band(destFlags, affiliation) <= 0
end
```

Used as: 
```lua
if event=="SPELL_INTERRUPT" and cast.by(ME) then
-- ...
end

-- ...

if ((
	event == "UNIT_DIED" or
	event == "UNIT_DESTROYED" or
	event == "UNIT_DESTROYED") and
	cast.on(ME) and
	cast.on(PET)
) then 

-- ...

if event=="SPELL_MISSED" and cast.notOn(ME) then
```

### CODE REVIEW

File: service/ErrorService.lua [hunk 1]
```diff
@@ -0,0 +1,41 @@
+local _, addon = ...
+
+function addon.ErrorsFrame_AddMessage(self, msg, ...)
+    local lowermsg = string.lower(msg)
+    local contains = addon.tableContains
+    local standardErrorMessages = addon.standardErrorMessages()
+
+    if (contains(standardErrorMessages, lowermsg)) then
```
Comment (jobackman) [inline service/ErrorService.lua:8]: ``
```suggestion
    if (addon.tableContains(standardErrorMessages, lowermsg)) then
```

File: SpellNotifications.lua [hunk 2]
```diff
@@ -40,304 +20,191 @@ function SpellNotifications_OnLoad(self)
...
+			destGUID==UnitGUID("arena1") or
+			destGUID==UnitGUID("arena2") or
+			destGUID==UnitGUID("arena3") or
+			destGUID==UnitGUID("arena4") or
+			destGUID==UnitGUID("arena5") or
+			destGUID==UnitGUID("party1") or
+			destGUID==UnitGUID("party2") or
+			destGUID==UnitGUID("party3") or
+			destGUID==UnitGUID("party4") or
+			destGUID==UnitGUID("party5")
```
Comment (jobackman) [inline SpellNotifications.lua:150]: We can probably redo this section into some kinda  `function addon.isRelevantTarget(GUID)` that runs the check, so we dont clutter this space so much

Comment (jobackman) [reply SpellNotifications.lua:150]: Then we'd have a pretty readable

```lua
if event == "SPELL_MISSED" and cast.by(ME) and isRelevantTarget(destGUID) then
-- ...
end
```

File: SpellNotifications.lua [hunk 3]
```diff
@@ -40,304 +20,191 @@ function SpellNotifications_OnLoad(self)
...
-				SpellNotifications_Print("Dispelled "..spellName..".","white","small") -- friendly target
-			else
-				SpellNotifications_Print("Dispelled "..spellName..".","yellow","small") -- enemy target
-			end
-		end
+	local _,class = UnitClass("player")
+	local size = addon.Sizes()
+	local color = addon.Colors()
+	local affiliation = addon.Affiliations()
+	local ME, FRIENDLY, PET = affiliation.MINE, affiliation.FRIENDLY, affiliation.PET
```
Comment (oscarwika) [inline SpellNotifications.lua:35]: Very nice 👍🏼

File: SpellNotifications.lua (hunk 2 above)
Comment (oscarwika) [reply SpellNotifications.lua:150]: Yeah good idea, hadn't really thought about how big this part was since everything was on the same line 😆

**Comment by jobackman** [comment]
Btw if you want a more direct contact, msg me or something, you can add my discord user ~<gone>~ in case I dont respond. Im probably more likely to to respond there cause Isuck at checking my email so I miss a lot of GH notifications 😆

File: data/MissTypes.lua [hunk 4]
```diff
@@ -8,6 +8,7 @@ function addon.MissTypes()
         ["PARRY"] = "parried",
         ["DODGE"] = "dodged",
         ["BLOCK"] = "blocked",
+        ["DEFLECT"] = "deflected",
```
Comment (jobackman) [inline data/MissTypes.lua:11]: Good catch!

File: SpellNotifications.lua [hunk 5]
```diff
@@ -40,304 +20,191 @@ function SpellNotifications_OnLoad(self)
...
+
+	local cast = {}
```
Comment (oscarwika) [inline SpellNotifications.lua:37]: I really like this part as well 👍🏼

File: SpellNotifications.lua [hunk 6]
```diff
@@ -40,304 +20,191 @@ function SpellNotifications_OnLoad(self)
...
+		addon.print("Interrupted "..string.lower(spellSchool)..".", color.GREEN, size.SMALL)
 	end
-	end
-
-
-
-
-	if (event=="SPELL_INTERRUPT") then
-		if bit.band(sourceFlags, COMBATLOG_OBJECT_AFFILIATION_MINE) > 0 then
-			local extraSchool = select(17, CombatLogGetCurrentEventInfo())
```
Comment (oscarwika) [inline SpellNotifications.lua:58]: Great to see code like this being deleted 🤣

**Review by oscarwika — APPROVED** [review]
I have no complaints, very nice work 😄 👍🏼 

Perhaps merge my PR into this branch before commiting to master?

**Comment by jobackman** [comment]
> Perhaps merge my PR into this branch before commiting to master?

Definitely 👍  I approved your PR when you I noticed that you pointed out it merges to this one 😋

**Comment by oscarwika** [comment]
> Btw if you want a more direct contact, msg me or something, you can add my discord user **LucyON#1510** in case I dont respond. Im probably more likely to to respond there cause Isuck at checking my email so I miss a lot of GH notifications 😆

Hahah, yeah sure. My GitHub notifications are also filtered to another folder as well so yeah 😆

**Comment by jobackman** [comment]
> Hahah, yeah sure. My GitHub notifications are also filtered to another folder as well so yeah 😆

image

😱

**Comment by oscarwika** [comment]
> > Hahah, yeah sure. My GitHub notifications are also filtered to another folder as well so yeah 😆
> 
> image
> 
> 😱

Hahah shit, I don't have that many 🤣

**Pull request merged**